### DISCUSSION TITLE
Please support URIs that are not HTTP/HTTPS in links

### DISCUSSION THREAD

**Discussion by fuzyll** [outline/outline#4514] [Ideas]
When I create a link that doesn't start with `http` or `https`, Outline will helpfully add it for me. Unfortunately, I sometimes don't want that because the link is using a different URI.

In my specific case, the [product my company develops](https://binary.ninja/) registers a custom URI handler with the OS so our application can open local or remote files with a link. When I add these URIs to Outline, it breaks them by putting `https://` in front. Being able to link to specific resources that users can open directly from Outline with other, non-browser applications would be a huge benefit to our workflow.

I *believe* the only change that would be required is to either not automatically add the `https://` in front, or give a user the option to prevent it.

**Comment by tommoor** [comment]
This should already work on getoutline.com and the latest pre-release build 👍

**Reply by fuzyll** [reply]
Thanks! It...doesn't work, though. 😞 

<img width="190" alt="image" src="https://user-images.githubusercontent.com/607452/206881498-c8059eea-615e-4eb3-b6ed-54d2696d1238.png">

I'm using the latest pre-release build in my deployment:

```
services:
  outline:
    depends_on:
      - "postgres"
      - "redis"
      - "minio"
    image: "outlinewiki/outline:0.67.0-pre.1"
```

...but, the server still says 0.66.2 when I go into settings:

<img width="139" alt="image" src="https://user-images.githubusercontent.com/607452/206881510-7e51c89b-8bed-4882-a85a-2d84fcd8dbb3.png">

I've confirmed that I am, in fact, running the correct container image:

```
9qrrcljjq30h   wiki_outline.1    outlinewiki/outline:0.67.0-pre.1   server2    Running         Running 8 minutes ago
```

Is there maybe something wrong with the image that got pushed to Docker Hub?

**Reply by tommoor** [reply]
Yea the image has the wrong version internally

**Reply by tommoor** [reply]
Have you tried recreating the link on that release?

**Reply by fuzyll** [reply]
Yes, that's what I was doing - trying to re-create the link.

Our URIs for our Enterprise product are of the form `binaryninja:enterprise-server:https://...`, which instructs our product to connect to an Enterprise server over `https` at a specific IP/hostname (and, optionally, port). When I enter a link of the form `[link name](binaryninja:enterprise-server:https://...)`, Outline helpfully puts an `https://` in front of it, which then breaks the link. It breaks it further if I try and remove the `https://` from the front after clicking on it and having the link target show up: It'll try to add _another_ `https://` on the front, resulting in `https://https://binaryninja:enterprise-server:https://...`.

**Reply by tommoor** [reply]
Unfortunately your url is not valid which is why it's being sanitized… the protocol cannot have a `:` character in it.

**Reply by fuzyll** [reply]
The URI is, indeed, valid. Another real-world example of this is [feed URIs](https://en.m.wikipedia.org/wiki/Feed_URI_scheme). I recall JDBC URLs functioning the same way, too.

**Reply by tommoor** [reply]
Feed URI's work fine on cloud now and will be fixed in next OSS release, there was a bug there that's been addressed. When I say that your url is not valid, this is what I mean:

![image](https://user-images.githubusercontent.com/380914/207180761-137e01f1-6f2a-4e4e-ad94-5c290bd99b18.png)

See how the protocol is not correctly parsed? That's because you can't have a `:` character in your protocol. Just a headsup.

**Reply by fuzyll** [reply]
Sorry, Tom, I’m still not understanding. That URI should be a completely valid RFC 2396 and 3986 URI, so far as I’m aware. The schema is `binaryninja` and the path is `enterprise-server:https://…`. Colons are valid characters inside a path, otherwise the aforementioned examples would also not be valid. URNs should be another example where multiple colons may intentionally exist (I am not asking for URN support, I am only mentioning it as an example).

I don’t understand how the protocol would be including a colon in the parsed-up version of the string you showed. I admit that I’m on travel and can’t verify conclusively right now, but I did check at least two online validators that both say the URI I provided is valid. Do you have a counter-example of a conformant library that says they are not?

**Reply by fuzyll** [reply]
Something I did not catch earlier was that, in the image you showed, you are trying to parse this URI as a *URL*. Our URIs are *not* valid URLs. This probably explains the issue.

Is there a better way I could describe the feature I want here? There are lots of URI parsers available in, for example, JavaScript (like [this one](https://www.npmjs.com/package/uri-js)) that could potentially be leveraged. I just want Outline to preserve the URI instead of trying to turn it into a URL.

I know this is not going to be a big priority for you. I just want to make sure what I want is being clearly communicated and understood.

Thanks!

**Reply by fuzyll** [reply]
Took a look at this tonight and I believe all that I would need/want is a way to short-circuit the logic in [`sanitizeUrl` from `shared/utils/urls.ts`](https://github.com/outline/outline/blob/main/shared/utils/urls.ts#L100). You already special-case some things like `mailto:` and `sms:`...I just need a way to either add my own special cases, or short-circuit that logic so that it won't prepend the `https://` anymore.

**Reply by tommoor** [reply]
If you can make a generic change in `urls.ts` that solves the issue you're seeing and does not regress any of the existing tests then I'd definitely be open to merging that PR.

**Reply by fuzyll** [reply]
Awesome, thanks! I'll see what I can do. Not super familiar with `jest` or `yarn`, but I see the unit test steps in the README so I'll try to set up a development environment and get to work on it soon.

**Comment by NamUKF** [comment]
I'd like to agree or additionally express my support for this feature / change of implementation for links. 

It'd nice if users were given more flexibility in terms of creating Links in general. For example we need want to share links to drives like: `D:\folder\file.txt` which currently not possible to share as a link, because the current link implementation automatically adds `https://` at the beginning of the link.
